filebeat/input/filestream/copytruncate_prospector.go

```diff
@@ -0,0 +1,357 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package filestream
+
+import (
+	"os"
+	"regexp"
+	"sort"
+	"strconv"
+	"time"
+
+	"github.com/urso/sderr"
+
+	loginp "github.com/elastic/beats/v7/filebeat/input/filestream/internal/input-logfile"
+	input "github.com/elastic/beats/v7/filebeat/input/v2"
+	"github.com/elastic/beats/v7/libbeat/logp"
+	"github.com/elastic/go-concert/unison"
+)
+
+const (
+	copyTruncateProspectorDebugKey = "copy_truncate_file_prospector"
+	copiedFileIdx                  = 0
+)
+
+// sorter is required for ordering rotated log files
+// The slice is ordered so the newest rotated file comes first.
+type sorter interface {
+	sort([]rotatedFileInfo)
+}
+
+// rotatedFileInfo stores the file information of a rotated file.
+type rotatedFileInfo struct {
+	path string
+	src  loginp.Source
+
+	ts  time.Time
+	idx int
+}
+
+func (f rotatedFileInfo) String() string {
+	return f.path
+}
+
+// rotatedFilestream includes the information of the original file
+// and its identifier, and the rotated file.
+type rotatedFilestream struct {
+	originalSrc loginp.Source
+	rotated     []rotatedFileInfo
+}
+
+func newRotatedFilestreams(cfg *copyTruncateConfig) *rotatedFilestreams {
+	var sorter sorter
+	sorter = newNumericSorter()
+	if cfg.DateFormat != "" {
+		sorter = &dateSorter{cfg.DateFormat}
+	}
+	return &rotatedFilestreams{
+		table:  make(map[string]*rotatedFilestream, 0),
+		count:  cfg.Count,
+		sorter: sorter,
+	}
+}
+
+// numericSorter sorts rotated log files that have a numeric suffix.
+// Example: apache.log.1, apache.log.2
+type numericSorter struct {
+	suffix *regexp.Regexp
+}
+
+func newNumericSorter() sorter {
+	return &numericSorter{
+		suffix: regexp.MustCompile("\\d*$"),
```

Comment:
store the compiled regex in a global (hidden) variable. Then Filebeat and CI will panic immediately if someone messed this one up.